pennylane/workflow/qnode.py

```diff
@@ -851,8 +898,8 @@ def _impl_call(self, *args, **kwargs) -> Result:
 
         return _to_qfunc_output_type(res, self._qfunc_output, tape.shots.has_partitioned_shots)
 
-    def __call__(self, *args, **kwargs) -> Result:
-        if "shots" in kwargs and qml.set_shots in self.transform_program:
+    def __call__(self, *args, **kwargs) -> qml.typing.Result:
+        if "shots" in kwargs and self._shots_override_device:
```

Comment:
Wait, what if the user does
```python
@qml.set_shots(10)
@qml.qnode(qml.device(..., shots=10))
def circuit():
    ...

circuit(shots=100)
```
I don't think the warning is raised here because `_shots_override_device` would be `False`

Reply:
> Test that the tape uses device shots only when qfunc has a shots kwarg
This is very confusing. It says "use device shots only when qfunc has a shots kwarg", then if qfunc doesn't have a shots kwarg what are you supposed to use? Is this actually expected?

Reply:
> > Test that the tape uses device shots only when qfunc has a shots kwarg
> 
> This is very confusing. It says "use device shots only when qfunc has a shots kwarg", then if qfunc doesn't have a shots kwarg what are you supposed to use? Is this actually expected?

Mind also sharing this in channel? My gut feeling this is also something deserve our attention at least some clarification and improvement phrasing.

Reply:
So with our new definition of precedence of `QNode._shots` I think this one should be fine? @astralcai